### ISSUE TITLE
[QWebEngine] qmake fails with Qt 5.6+

### ISSUE THREAD

**Issue by u-235** [gyunaev/kchmviewer#8]
# [QWebEngine] qmake fails with Qt 5.15

Qmake fails when building from `RELEASE_8_0` with Qt 5.15.

## Environment

- **CPU** Intel(R) Pentium(R) Dual CPU E2180 @ 2.00GHz
- **RAM** 3GB
- **OS** Linux Mint 18.1 Serena with kernel Linux 4.4.0-53-generic i686 GNU/Linux
- **DE** MATE Desktop Environment 1.16.2
- **Compiler** g++ (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
- **Qt 5.15.2** local build with QWebEngine support
- **Libraries** libchm1 libchm-dev libzip4 libzip-dev

## Run qmake

```
base$ git clone https://github.com/gyunaev/kchmviewer.git
base$ mkdir build
base$ cd build
base/build$ qmake -r ../kchmviewer
```

Qmake output:

```
Reading $base$/kchmviewer/lib/lib.pro [$base$/build/lib]
 Reading $base$/kchmviewer/lib/libebook/libebook.pro [$base$/build/lib/libebook]
Reading $base$/kchmviewer/src/src.pro [$base$/build/src]
Project ERROR: Unknown module(s) in QT: webkit webkitwidgets
```

**Comment by gyunaev** [comment]
Gonna leave this open until someone bothers install Ubuntu and fix it.

**Comment by gyunaev** [comment]
You also submitted two different build errors in the same environment. Is this one still current? Did you invoke the right qmake?

`base/build$ qmake -v` output?

**Comment by u-235** [comment]
> `base/build$ qmake -v` output?

```
QMake version 3.1
Using Qt version 5.15.2 in /tools/sdk/Qt/5.15.2/gcc/lib
```

**Comment by gyunaev** [comment]
You also submitted two different build errors in the same environment. Is this one still current?

**Comment by u-235** [comment]
It is build after meaningless changes in src/src.pro.

```diff
diff --git a/src/src.pro b/src/src.pro
index f502259..6047d60 100644
--- a/src/src.pro
+++ b/src/src.pro
@@ -40,7 +40,6 @@ SOURCES += config.cpp \
     textencodings.cpp \
     treeitem_toc.cpp \
     treeitem_index.cpp
-LIBS += -lchm -lzip
 TARGET = ../bin/kchmviewer
 CONFIG += threads \
     warn_on \
@@ -60,11 +59,9 @@ FORMS += tab_bookmarks.ui \
     toolbareditor.ui
 RESOURCES += resources/images.qrc
 
-QT += webkit \
-	xml \
+QT += xml \
     network \
     widgets \
-    webkitwidgets \
     printsupport
 
 linux-g++*:{
@@ -112,6 +109,8 @@ unix:!macx: {
     POST_TARGETDEPS += ../lib/libebook/libebook.a
 }
 
+LIBS += -lchm -lzip
+
 greaterThan(QT_MAJOR_VERSION, 4) {
     # Qt 5
     greaterThan(QT_MINOR_VERSION, 5) {

```

**Comment by gyunaev** [comment]
Can you please post the gcc linker command before the changes and after?

**Comment by u-235** [comment]
Yes, of course. But I can't do it until monday.

**Comment by gyunaev** [comment]
No worries, we aren't in hurry. Thank you for your help.

**Comment by u-235** [comment]
Patch to fix qmake error.

```diff
diff --git a/src/src.pro b/src/src.pro
index f502259..000a7e4 100644
--- a/src/src.pro
+++ b/src/src.pro
@@ -60,11 +60,9 @@ FORMS += tab_bookmarks.ui \
     toolbareditor.ui
 RESOURCES += resources/images.qrc
 
-QT += webkit \
-	xml \
+QT += xml \
     network \
     widgets \
-    webkitwidgets \
     printsupport
 
 linux-g++*:{

```

Patch to fix linker errors.

```diff
diff --git a/src/src0.pro b/src/src.pro
index 000a7e4..6047d60 100644
--- a/src/src0.pro
+++ b/src/src.pro
@@ -40,7 +40,6 @@ SOURCES += config.cpp \
     textencodings.cpp \
     treeitem_toc.cpp \
     treeitem_index.cpp
-LIBS += -lchm -lzip
 TARGET = ../bin/kchmviewer
 CONFIG += threads \
     warn_on \
@@ -110,6 +109,8 @@ unix:!macx: {
     POST_TARGETDEPS += ../lib/libebook/libebook.a
 }
 
+LIBS += -lchm -lzip
+
 greaterThan(QT_MAJOR_VERSION, 4) {
     # Qt 5
     greaterThan(QT_MINOR_VERSION, 5) {

```

Linker command line before and after the second patch. 

```sh
# before
g++ -Wl,-O1 -Wl,-rpath,/tools/sdk/Qt/5.15.2/gcc/lib -Wl,-rpath-link,/tools/sdk/Qt/5.15.2/gcc/lib -o ../bin/kchmviewer config.o dialog_chooseurlfromlist.o dialog_setup.o kde-qt.o main.o mainwindow.o recentfiles.o settings.o tab_bookmarks.o tab_contents.o tab_index.o tab_search.o viewwindowmgr.o navigationpanel.o checknewversion.o toolbarmanager.o toolbareditor.o textencodings.o treeitem_toc.o treeitem_index.o dbus_interface.o viewwindow_webengine.o dataprovider_qwebengine.o qrc_images.o moc_dialog_chooseurlfromlist.o moc_dialog_setup.o moc_mainwindow.o moc_recentfiles.o moc_tab_bookmarks.o moc_tab_contents.o moc_tab_index.o moc_tab_search.o moc_viewwindowmgr.o moc_navigationpanel.o moc_checknewversion.o moc_toolbarmanager.o moc_toolbareditor.o moc_dbus_interface.o moc_viewwindow_webengine.o   -lchm -lzip -lX11 ../lib/libebook/libebook.a libQt5WebEngineWidgets.so libQt5PrintSupport.so libQt5Widgets.so libQt5WebEngine.so libQt5WebEngineCore.so libQt5Quick.so libQt5Gui.so libQt5Xml.so libQt5QmlModels.so libQt5WebChannel.so libQt5Qml.so libQt5Network.so libQt5DBus.so libQt5Positioning.so libQt5Core.so -lGL -lpthread
# after
g++ -Wl,-O1 -Wl,-rpath,/tools/sdk/Qt/5.15.2/gcc/lib -Wl,-rpath-link,/tools/sdk/Qt/5.15.2/gcc/lib -o ../bin/kchmviewer config.o dialog_chooseurlfromlist.o dialog_setup.o kde-qt.o main.o mainwindow.o recentfiles.o settings.o tab_bookmarks.o tab_contents.o tab_index.o tab_search.o viewwindowmgr.o navigationpanel.o checknewversion.o toolbarmanager.o toolbareditor.o textencodings.o treeitem_toc.o treeitem_index.o dbus_interface.o viewwindow_webengine.o dataprovider_qwebengine.o qrc_images.o moc_dialog_chooseurlfromlist.o moc_dialog_setup.o moc_mainwindow.o moc_recentfiles.o moc_tab_bookmarks.o moc_tab_contents.o moc_tab_index.o moc_tab_search.o moc_viewwindowmgr.o moc_navigationpanel.o moc_checknewversion.o moc_toolbarmanager.o moc_toolbareditor.o moc_dbus_interface.o moc_viewwindow_webengine.o   -lX11 ../lib/libebook/libebook.a -lchm -lzip libQt5WebEngineWidgets.so libQt5PrintSupport.so libQt5Widgets.so libQt5WebEngine.so libQt5WebEngineCore.so libQt5Quick.so libQt5Gui.so libQt5Xml.so libQt5QmlModels.so libQt5WebChannel.so libQt5Qml.so libQt5Network.so libQt5DBus.so libQt5Positioning.so libQt5Core.so -lGL -lpthread
```

**Comment by gyunaev** [comment]
Fixed in 9ac73e7..e3b09ed

Indeed it does seem to make a difference. Thank you for reporting this and for pushing through. Newer linkers (gcc 7+) do not need this, because they do multiple passes. But for older gcc this seem to be needed.
